sway/commands/bind.c

```diff
@@ -649,101 +646,144 @@ void seat_execute_command(struct sway_seat *seat, struct sway_binding *binding)
-			find_keycode, &matches);
-	return matches;
+xkb_keycode_t **cartesian_product(list_t **sets, int n) {
+	int total_combinations = 1;
+	for (int i = 0; i < n; ++i) {
+		total_combinations *= sets[i]->length;
+	}
+		
+	// Allocate memory for the current_result
+	int result_size = 0;
+	xkb_keycode_t **result = malloc(total_combinations * sizeof(xkb_keycode_t*));
+	xkb_keycode_t *current_product = malloc(n * sizeof(xkb_keycode_t));
+	cartesian_product_helper(sets, n, result, &result_size, current_product, 0);
+	free(current_product);
+
+	return result;
 }
 
+/*
+ * Convert keysyms to keycodes for --to-code. If any keysym does not have at
+ * least 1 keycode: halt conversion and unset BINDING_CODE. Assumes
+ * identify_key() read in binding->keys in keysym format vs keycode format
+ */
 bool translate_binding(struct sway_binding *binding) {
-	if ((binding->flags & BINDING_CODE) == 0) {
-		return true;
-	}
 
-	switch (binding->type) {
-	// a bindsym to translate
-	case BINDING_KEYSYM:
-		binding->syms = binding->keys;
-		binding->keys = create_list();
-		break;
-	// a bindsym to re-translate
-	case BINDING_KEYCODE:
-		list_free_items_and_destroy(binding->keys);
-		binding->keys = create_list();
-		break;
-	default:
+	if (binding->type != BINDING_KEYSYM ||
+		(binding->flags & BINDING_CODE) == 0) {
 		return true;
 	}
 
-	for (int i = 0; i < binding->syms->length; ++i) {
-		xkb_keysym_t *keysym = binding->syms->items[i];
-		struct keycode_matches matches = get_keycode_for_keysym(*keysym);
+	int num_syms = binding->keys->length;
+	list_t **sym2code = malloc(num_syms * sizeof(list_t*)); 
+	// Collect all keycodes for all keysyms
+	for (int i = 0; i < num_syms; i++) {
+		struct keycode_matches matches = {
+			.keysym = (xkb_keysym_t*)binding->keys->items[i],
+			.keycodes = create_list(),
+			.error = false
+		};
 
-		if (matches.count != 1) {
-			sway_log(SWAY_INFO, "Unable to convert keysym %" PRIu32 " into"
-					" a single keycode (found %d matches)",
-					*keysym, matches.count);
+		xkb_keymap_key_for_each(
+				xkb_state_get_keymap(config->keysym_translation_state),
+				add_matching_keycodes, &matches);
+
+		if (matches.error) {
+			sway_log(SWAY_ERROR, "Failed to allocate memory for keycodes while iterating for keysym %" PRIu32, *matches.keysym);
 			goto error;
 		}
 
-		xkb_keycode_t *keycode = malloc(sizeof(xkb_keycode_t));
-		if (!keycode) {
-			sway_log(SWAY_ERROR, "Unable to allocate memory for a keycode");
-			goto error;
+		if (matches.keycodes->length == 0) {
+		   sway_log(SWAY_INFO, "Unable to convert keysym %" PRIu32 " into"
+					  "any keycodes", *matches.keysym);
+		   goto error;
 		}
 
-		*keycode = matches.keycode;
-		list_add(binding->keys, keycode);
+		sym2code[i] = matches.keycodes;
 	}
 
-	list_qsort(binding->keys, key_qsort_cmp);
-	binding->type = BINDING_KEYCODE;
+	// If any keycode maps to more than one keysym, use all combinations.
+	xkb_keycode_t **combos = cartesian_product(sym2code, num_syms);
```

Comment:
So the implemented strategy is to collect all codes in advance in a list of lists, and then create a new list-of-list containing the Cartesian product of all the codes, and then finally create bindings using them. I can't help but feel that this is a bit too complicated, with a lot of cleanup needed.

How about a recursive strategy, where we just register the keybindings as we iterate? Something along the lines of this totally untested psuedocode:

```c

struct translate_ctx {
        struct xkb_keymap *keymap;
        struct sway_binding *binding;
        xkb_keycode_t *codes;
        int depth;
        bool error;
};

static void recursive_translate(struct xkb_keymap *keymap, xkb_keycode_t keycode, void *data) {
        struct translate_ctx *ctx = data;
        ctx->codes[ctx->depth] = keycode;

        if (depth == binding->keys->length) {
                // We have a pick for each symbol, so add it.
                if (!add_binding(ctx->binding, ctx->codes)) {
                        ctx->error = true;
                }
                return;
        }

        ctx->depth++;
        xkb_keymap_key_for_each(ctx->keymap, recursive_translate, ctx);
        ctx->depth--;
}

bool translate_binding(struct sway_binding *binding) {
        xkb_keycode_t codes[binding->keys->length] = { 0 };
        struct translate_ctx ctx = {
                .keymap = xkb_state_get_keymap(config->keysym_translation_state),
                .binding = binding,
                .codes = codes,
                .depth = 0,
        };
        xkb_keymap_key_for_each(ctx->keymap, recursive_translate, ctx);
        return !ctx.error;
}
```

Reply:
Doesn't this just choose the first 3 keycodes in the keymap for all the bindings? I'm not sure exactly what the output of that is but it doesn't look like it spits out a cartesian map of the keysyms.

Reply:
In fact, correct me if I'm wrong, but I don't think it is mathematically possible to map out a cartesian combination of Sets A x B x C while also adding elements to those sets at the same time in the same "process"

Reply:
Let's say we 3 symbols (A, B), and each of them have 2 keycodes (A1, A2, etc.).

1. The first iterator is called at depth 0, which is looking for all codes matching symbol A.
2. The first iterator finds `A1`, and stores it at `keycodes[0]`. It then calls the iterator for the next level.
3. The second iterator finds `B1`, and stores it at `keycodes[1]`. It sees that we reached the last symbol, and stores the binding for `[A1, B1]`.
4. The second iterator finds `B2`, and stores it at `keycodes[1]. It sees that we reached the last symbol, and stores the binding for `[A1, B2]`.
5. The second iterator finishes.
6. The first iterator finds `A2`, and stores it at `keycodes[0]`. It then calls the iterator for the next level.
7. Second goes through the paces again, but with `A2` at the first position.
8. The first iterator finishes, at which point we have stored: `[A1, B1], [A1, B2], [A2, B1], [A2, B2]`.

This should scale to any number of symbols, although at the cost of iterating over the keymap a lot more than using the cartesian product approach. However, it is at least to me a lot simpler.

Reply:
```
if (keysym != binding->syms->items[ctx->depth]) {
            return;
        }
```
Ok now I see what you're going for with this addition. 
I think that should work. I'll try it out later and report back.

Reply:
I couldn't get your algorithm to not segfault. Can you get it to work?

Reply:
Yeah, I have a local patch with the recursive approach that works with bind and unbind, but haven't changed the retranslation logic.

I suspect that the easiest solution is to find and remove all "derived" bindings and retranslate them from their original description (which will either have to be extracted and deduplicated, or stored separately).

Reply:
Does your patch work on translating `bindsym --to-code Print+a`?

Reply:
The easiest solution could be to merge my patch that works.

Reply:
> The easiest solution could be to merge my patch that works.

Your patch also does not handle retranslation of existing bindings correctly - it removed that logic.

Reply:
If you're going to merge that commit in, should I just close this PR? And timeline?